### PR TITLE
[12.0][ADD] delivery_state: Delivery status tracking

### PR DESCRIPTION
**Pull request by rlizana**
This module adds additional functions that will be needed for the carrier developments. It provides fields to be able to contemplate the tracking states and also adds a global field so it can have generic states in addition to the ones carrier gives us.

SPANISH OCA Days 2020

### CODE REVIEW

File: delivery_carrier_state/i18n/es.po [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+"Language-Team: \n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: \n"
+"Plural-Forms: \n"
+
+#. module: delivery_carrier_state
+#: selection:stock.picking,delivery_state:0
+msgid "Canceled shipment"
+msgstr "Envio cancelado"
```
Comment (azucenatrey) [inline delivery_carrier_state/i18n/es.po]: The accent mark in "Envío" is missing.

File: delivery_carrier_state/i18n/es.po [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+#. module: delivery_carrier_state
+#: selection:stock.picking,delivery_state:0
+msgid "incidence"
+msgstr "incidencia"
+
+#. module: delivery_carrier_state
+#: selection:stock.picking,delivery_state:0
+msgid "shipping recorded in carrier"
+msgstr "envio registrado en el transportista"
```
Comment (azucenatrey) [inline delivery_carrier_state/i18n/es.po]: The accent mark in "envío" is missing.

File: delivery_carrier_state/models/delivery_carrier.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = 'delivery.carrier'
+
+    def send_shipping(self, pickings):
+        res = super().send_shipping(pickings)
+        pickings.write({
+            'delivery_state': 'shipping_recorded_in_carrier',
+            'date_shipped': fields.Date.today()})
```
Comment (azucenatrey) [inline delivery_carrier_state/models/delivery_carrier.py]: I think it's more "clean code" like so:

```
            'date_shipped': fields.Date.today(),
        })
```

File: delivery_carrier_state/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        readonly=True,
+        index=True,
+    )
+    tracking_state_history = fields.Text(
+        string='Tracking state history',
+        readonly=True,
+    )
+    delivery_state = fields.Selection(
+        selection=[
+            ('shipping_recorded_in_carrier', 'shipping recorded in carrier'),
```
Comment (azucenatrey) [inline delivery_carrier_state/models/stock_picking.py]: The label should start with a capital letter: 

```
('shipping_recorded_in_carrier', 'Shipping recorded in carrier'),
```

File: delivery_carrier_state/models/stock_picking.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+            ('in_transit', 'In transit'),
+            ('canceled_shipment', 'Canceled shipment'),
+            ('incidence', 'incidence'),
```
Comment (azucenatrey) [inline delivery_carrier_state/models/stock_picking.py]: The label should start with a capital letter: 

```
('incidence', 'Incidence'),
```

File: delivery_carrier_state/views/stock_picking.xml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+    <record id="view_picking_withcarrier_out_form" model="ir.ui.view">
+        <field name="model">stock.picking</field>
```
Comment (azucenatrey) [inline delivery_carrier_state/views/stock_picking.xml]: The "name" field is not required in views, it can be omitted.

Comment (rlizana) [reply delivery_carrier_state/views/stock_picking.xml]: Ups!...
You got confused, it's the model :yum:

Comment (azucenatrey) [reply delivery_carrier_state/views/stock_picking.xml]: Sorry i have seen it wrong.

File: delivery_carrier_state/README.rst [hunk 7]
```diff
@@ -0,0 +1,102 @@
+======================
+Delivery carrier state
+======================
+
+.. !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+   !! This file is generated by oca-gen-addon-readme !!
+   !! changes will be overwritten.                   !!
+   !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Comment (chienandalu) [inline delivery_carrier_state/README.rst]: Just use readme parts that generate the readme automatically :sweat_smile: 
Put the fragments you need in the readme folder and OCA bot will do the rest (or you can generate it in advance with https://github.com/OCA/maintainer-tools#readme-generator wich prevents errors). Estructure template:
https://github.com/OCA/maintainer-tools/tree/master/template/module/readme

File: delivery_carrier_state/__init__.py [hunk 8]
```diff
@@ -0,0 +1,4 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
Comment (chienandalu) [inline delivery_carrier_state/__init__.py]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

File: delivery_carrier_state/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU Affero General Public License for more details.
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
```
Comment (chienandalu) [inline delivery_carrier_state/__manifest__.py]:
```suggestion
# Copyright 2020 Trey, Kilobytes de Soluciones
# Copyright 2020 FactorLibre
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

File: delivery_carrier_state/models/__init__.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
Comment (chienandalu) [inline delivery_carrier_state/models/__init__.py]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

File: delivery_carrier_state/models/delivery_carrier.py [hunk 11]
```diff
@@ -0,0 +1,16 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
Comment (chienandalu) [inline delivery_carrier_state/models/delivery_carrier.py]:
```suggestion
# Copyright 2020 Trey, Kilobytes de Soluciones
# Copyright 2020 FactorLibre
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

File: delivery_carrier_state/models/stock_picking.py [hunk 12]
```diff
@@ -0,0 +1,51 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
Comment (chienandalu) [inline delivery_carrier_state/models/stock_picking.py]:
```suggestion
# Copyright 2020 Trey, Kilobytes de Soluciones
# Copyright 2020 FactorLibre
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

File: delivery_carrier_state/tests/__init__.py [hunk 13]
```diff
@@ -0,0 +1,4 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
Comment (chienandalu) [inline delivery_carrier_state/tests/__init__.py]:
```suggestion
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

File: delivery_carrier_state/tests/test_delivery_carrier_state.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
+###############################################################################
+# For copyright and license notices, see __manifest__.py file in root directory
+###############################################################################
```
Comment (chienandalu) [inline delivery_carrier_state/tests/test_delivery_carrier_state.py]:
```suggestion
# Copyright 2020 Trey, Kilobytes de Soluciones
# Copyright 2020 FactorLibre
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
```

File: delivery_carrier_state/models/stock_picking.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+        readonly=True,
+    )
+
+    def tracking_state_update(self):
+        for picking in self.filtered(lambda p: p.delivery_type):
+            method = '%s_tracking_state_update' % picking.delivery_type
+            if hasattr(picking.carrier_id, method):
+                getattr(picking.carrier_id, method)(picking)
+
+    def cancel_shipment(self, picking):
```
Comment (chienandalu) [inline delivery_carrier_state/models/stock_picking.py]: Why not doing it in the `cancel_shipment` method of  `delivery.carrier`? It should return the recordset of cancelled of pickings with cancelled deliveries: https://github.com/odoo/odoo/blob/12.0/addons/delivery/models/delivery_carrier.py#L158.

Indeed, you could have a delivery that can't be cancelled (i.e.: it's already delivered) wich state shouldn't be changed.

File: delivery_carrier_state/models/delivery_carrier.py [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+from odoo import fields, models
+
+
+class DeliveryCarrier(models.Model):
+    _inherit = 'delivery.carrier'
+
+    def send_shipping(self, pickings):
+        res = super().send_shipping(pickings)
+        pickings.write({
+            'delivery_state': 'shipping_recorded_in_carrier',
```
Comment (chienandalu) [inline delivery_carrier_state/models/delivery_carrier.py]: What if there are pickings not being recorded? Maybe this should be in a separate method and do it only by default for the core methods. What do you think about it?

File: delivery_carrier_state/models/stock_picking.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+            ('warehouse_delivered', 'Warehouse delivered'),
+        ],
+        string='Carrier State',
+        track_visibility='onchange',
+        readonly=True,
+    )
+
+    def tracking_state_update(self):
+        for picking in self.filtered(lambda p: p.delivery_type):
+            method = '%s_tracking_state_update' % picking.delivery_type
```
Comment (chienandalu) [inline delivery_carrier_state/models/stock_picking.py]: I think provider methods should belong to `delivery.carrier` like the core convention states: https://github.com/odoo/odoo/blob/12.0/addons/delivery/models/delivery_carrier.py#L19-L31

**Review by chienandalu — CHANGES_REQUESTED** [review]
Thanks @rlizana some remarks:

File: delivery_carrier_state/README.rst (hunk 7 above)
Comment (rlizana) [reply delivery_carrier_state/README.rst]: @chienandalu Thanks for the info, I do it this way

File: delivery_carrier_state/models/stock_picking.py (hunk 15 above)
Comment (rlizana) [reply delivery_carrier_state/models/stock_picking.py]: You're right, I change it

File: delivery_carrier_state/models/delivery_carrier.py (hunk 16 above)
Comment (rlizana) [reply delivery_carrier_state/models/delivery_carrier.py]: If has not been registered must thrown an exception.

File: delivery_carrier_state/models/stock_picking.py (hunk 17 above)
Comment (rlizana) [reply delivery_carrier_state/models/stock_picking.py]: it is in the delivery.carrier where the methods must be, no in stock.picking, revise it

File: delivery_carrier_state/models/delivery_carrier.py (hunk 16 above)
Comment (chienandalu) [reply delivery_carrier_state/models/delivery_carrier.py]: Ok :+1:

**Comment by rlizana** [comment]
@chienandalu can you review?

**Comment by pedrobaeza** [comment]
Hi, sorry for the delay in the review, although most of the work was already done. I have directly push to your branch for GTD with this with this little changes - I hope you don't mind -:

- Squash all the commits, as they come from the same author and don't add meaningful split in the final merge.
- Remove the copyrights in `__init__.py` files, as they shouldn't be put there.
- Added tracking to `delivery_state` field for having more verbose changes.
- Pre-generated README.rst

With that, I proceed to merge for unblocking the dependencies:

/ocabot merge nobump

**Comment by pedrobaeza** [comment]
I have just detected an RST incorrect syntax, so fixed also

/ocabot merge nobump

**Pull request merged**